<img src='imgs/logo.jpg'>


# Local version of [PhaBOX2](https://phage.ee.cityu.edu.hk) web server

[![PhaBOX Server](https://img.shields.io/badge/PhaBOX-Webserver-brightgreen)](http://phage.ee.cityu.edu.hk/)
[![GitHub License](https://img.shields.io/github/license/KennthShang/PhaBOX)](https://github.com/KennthShang/PhaBOX/blob/main/LICENSE.md)
[![BioConda Install](https://img.shields.io/conda/dn/bioconda/phabox.svg?style=flag&label=BioConda%20install)](https://anaconda.org/bioconda/phabox) 
[![PhaBOXv1](https://img.shields.io/static/v1.svg?label=PhaBOX_v1&message=bioadv/vbad101&color=blue)](https://doi.org/10.1093/bioadv/vbad101)
[![PhaMer](https://img.shields.io/static/v1.svg?label=PhaMer&message=bib/bbac258&color=blue)](https://doi.org/10.1093/bib/bbac258)
[![PhaGCN](https://img.shields.io/static/v1.svg?label=PhaGCN&message=bioinformatics/btab293&color=blue)](https://doi.org/10.1093/bioinformatics/btab293)
[![PhaTYP](https://img.shields.io/static/v1.svg?label=PhaTYP&message=bib/bbac487&color=blue)](https://doi.org/10.1093/bib/bbac487)
[![CHERRY](https://img.shields.io/static/v1.svg?label=CHERRY&message=bib/bbac182&color=blue)](https://doi.org/10.1093/bib/bbac182)
[![PhaVIP](https://img.shields.io/static/v1.svg?label=CHERRY&message=bioinformatics/btad229&color=blue)](https://doi.org/10.1093/bioinformatics/btad229)

This is the source code of our website [PhaBOX2](https://phage.ee.cityu.edu.hk). 

For large-scale sequencing data, please download this local version and use more threads to speed up the program.

<a name="news"></a>
## âŒ›ï¸&nbsp; News

PhaBOX has now been upgraded to the 2.0 version with faster speed!

There are some major components, including:

  ğŸ‰ Generalized for all kinds of viruses; more than just bacteriophage

  ğŸ‰ Virus identification (PhaMer)

  ğŸ‰ Taxonomy classification (PhaGCN)

  ğŸ‰ Host prediction (CHERRY/HostG)

  ğŸ‰ Lifestyle prediction (PhaTYP)

  ğŸ‰ Contamination/provirus detection

  ğŸ‰ vOTU grouping

  ğŸ‰ Phylogenetic tree based on marker genes

  ğŸ‰ Viral protein annotation

  ğŸ‰ All the databases are updated to the latest ICTV 2024 release

If you have any more suggestions, feel free to let me know! We consider long-term maintenance PhaBOX and adding modules according to your needs


You can post an issue or directly email me (jiayushang@cuhk.edu.hk). We welcome any suggestions.

<a name="quick"></a>
## ğŸš€&nbsp; Quick Start
> [!IMPORTANT]
> If you are a new user, please check our [WIKI](https://github.com/KennthShang/PhaBOX/wiki) page. We provide a tutorial to help you get started quickly and understand how to use PhaBOX2. We hope you will enjoy it!
>
> If you only want to analyze the results for Bacteriophages (as version 1), please check the **Prokarytic viruses** columns in the PhaGCN's results.


If you are familiar with the PhaBOX2, please check our [Update log](https://github.com/KennthShang/PhaBOX/wiki/Update-logs). We may have some updates to the program to make it more useful. If you want to use the latest version, please also [upgrade your PhaBOX2](https://github.com/KennthShang/PhaBOX/wiki#upgrading-phabox)


<a name="license"></a>

## ğŸ“˜&nbsp; License
The PhaBOX pipelines are released under the terms of the [Academic Free License v3.0 License](https://choosealicense.com/licenses/afl-3.0/).

